platform-sdk/docs/proposals/consensus-layer/Consensus-Layer.md

```diff
@@ -0,0 +1,709 @@
+the new roster. The Hashgraph module will decide in what round the new roster becomes active. Let us define `N` as the
+number of rounds from the latest consensus round at which the candidate roster will become active. This number will be
+configured with the same value for all nodes in the network.
+
+For example, suppose Execution has just finished handling round 10, and has buffered up round 11. The most recent round
+that the Hashgraph module produced was round 11. Suppose `N` is 10. If the call to `nextRound` from Execution at the
+end of round 10 were to supply a new Roster (`R`). The Hashgraph module will assign `R` to become active as of round 22
+(`latest_consensus_round + N + 1`). It will then work on producing consensus round 12. Once produced, metadata
+indicating that `R` will be used for rounds starting at 22 is passed to Gossip, Event Intake, Event Creator, and
+Execution. The roster used for round 22 may also be included in this metadata.
+
+There is a critical relationship between the value of N, the latency for adopting a new roster, and the number of future
+events held in memory on the node. Remember that each event has a birth-round. Let us define another variable called
+`Z`, such that `Z <= N`. Any event with `birth_round > latest_consensus_round + z` is considered a "far future" event,
+and will be dropped by the birth-round filtering logic in Event Intake. Any event with
+`birth_round > latest_consensus_round` is just a "future" event.
+
+The smaller the value of `Z`, the fewer events are held in memory on the node. However, events with birth-rounds just
+larger than `Z` will be dropped and have to be retrieved on demand with sync gossip, rather than through the broadcast
+tree. Sync is slower and more laborious than broadcast gossip, so the efficiency of the node will go down with low
+values of `Z`. Larger values of `Z` means more events in memory, but it also means more "smoothing" in gossip and
+in handling any performance spikes.
+
+On the other hand, the larger the value of `N`, the larger the latency between the round we know about a new roster, and
+the round at which it can be actually used. Ideally, the value of `N` would be small, like 3, but we may find that 3 is
+too small a number for `Z`.
+
+Each node may select its own value of `Z`, so long as it is less than or equal to `N`. But all nodes must use the same
+value for `N`, or they will ISS since they will assign different rosters to different rounds. The value of `N` is not
+defined here, but will need some investigation. Unfortunately, this number appears to require "just being chosen" and
+cannot be deterministically dynamically computed based on network conditions.
+
+Note: when the roster is applied to some future round `latest_consensus_round + N + 1`, if such a round already exists
+in the hashgraph, then that round and any future rounds must be removed and the events reapplied in order given the new
+roster.
```

Comment:
How can such a future round be produced if above we require every node to have the same value of `N`? Wouldn't they drop all such future events and never produce such a future round?

Also, is it technically possible to "remove rounds" and "reapply events" if that future round already exists in the hashgraph?

Reply:
This is internal consensus algorithm stuff. "Future rounds", in this context, are never visible outside of the consensus algorithm.

> Also, is it technically possible to "remove rounds" and "reapply events" if that future round already exists in the hashgraph?

Yes, Lazar implemented this. It's a required prerequisite for dynamic address books.

Reply:
Interesting. Unsure if it's a great idea though.

Getting back to the "future rounds" thing, I still don't understand how this could happen given that all nodes have the same value of `N` and hence they drop events that refer to future rounds.

It seems that this can only happen if a single node is very slow and is unable to keep up with the rest of the network. The rest of the network may have produced those `N` rounds already while the slow node is still working on an older one. Then this slow node will encounter the situation described in this paragraph, but this won't affect the rest of the network.

If my understanding above is correct, then this situation looks no different from when a node is slow and has to go through catching-up and/or reconnect as described above in the document. And if it's so, then I don't think this situation should be handled in the roster change implementation as this seems to unnecessarily complicate it given that we have another mechanism for nodes to catch up.

Reply:
> Interesting. Unsure if it's a great idea though.

I suggest you take it up with Leemon. He's the primary architect of this part of the algorithm.

Reply:
I'd like to clarify the "future rounds" thing first.

Reply:
Here is my understanding: 
* Roster A is the active roster.  
* Roster B becomes active on round X+2
* Latest Consensus Round: X
* Pending Consensus Round: X+1
* There are "rounds created" meta data on the events in consensus that indicate Round X+1, X+2, X+3, etc.  
     * The future rounds are > X+1
* The events with meta-data of round X+1 will have their round number committed and the events will be removed from consensus and put into the round object when that round comes to consensus. 
* The rest of the events have their meta-data erased.  All those future "round created" values, X+2, X+3, etc. go away. 
* At the start of pending consensus round X+2, the new roster is adopted. 
* The events that remain in consensus get new meta-data for "round created".  Some events get X+2, some X+3, X+4, etc.  
     * The future rounds are any > X+2
     * It is important to note here that events that were previously labeled with a future rounds when X+1 was the pending consensus round may get a different future round assigned in the calculation of pending consensus round X+2.   This is especially true due to the new address book being used for pending consensus round X+2

Reply:
In my example, the roster that takes effect on round X+2 would have been set to take effect N rounds earlier on round X+2-N.    Once a new roster is set to take effect, I'm talking about the behavior of consensus calculations on the round right before and on the round that the new roster takes effect.

Reply:
The paragraph that I've quoted to start this comment thread refers to a situation where a node has just discovered that there's a new roster to be applied to a round `latest_consensus_round + N + 1`.

If, as you say, the new roster has already been supplied at round `X+2-N` to be applied at round `X+2` and the latest consensus round is currently `X`, then all the nodes have processed the round `X+2-N` and already saw the new roster to be applied at `X+2`, and so your description doesn't apply to the situation quoted in that specific paragraph.

What you're saying is correct overall, but it doesn't describe the specific scenario described in this document that I've commented on. So this question is still open.

Reply:
> How can such a future round be produced if above we require every node to have the same value of N? Wouldn't they drop all such future events and never produce such a future round?

I don't understand the initial question of the thread.

@anthony-swirldslabs Have subsequent discussions cleared this up for you?

Reply:
Yes, that's correct. IIRC, in that meeting we established that a node will always learn about a new future roster at least N rounds in advance. This means that there'll never be a case where a node suddenly sees a new roster to be applied earlier than in N rounds from the current round. And therefore, the special edge case described in the quoted paragraph simply cannot happen. Meaning that the code responsible for the activation of the new roster will not need to "reapply events", at least not because of a roster that has to be suddenly applied in less than N rounds from the current round.

So we need to remove the quoted paragraph from the document because this specific situation simply cannot happen, and then this conversation can be considered resolved.

Reply:
The paragraph has already been removed - resolving this thread.